File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let time = 0;
    let bridge = Array(bridge_length).fill(0);
    let bridge_sum = 0;

    while (truck_weights.length > 0 || bridge_sum > 0) {
        // 시간 증가
        time++;

        // 다리에서 나가는 트럭 처리
        bridge_sum -= bridge.shift();

        // 새 트럭이 다리에 올라갈 수 있는지 확인
        if (truck_weights.length > 0 && bridge_sum + truck_weights[0] <= weight) {
            let truck = truck_weights.shift();
            bridge.push(truck);
            bridge_sum += truck;
        } else {
            bridge.push(0);
        }
    }

    return time;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    answer = 0
    bridge = deque([0] * bridge_length)
    trucks = deque(truck_weights)
    bridge_weight = 0
    total_trucks = len(truck_weights)
    passed = 0

    while passed < total_trucks:
        answer += 1
        out = bridge.popleft()
        if out != 0:
            passed += 1
        bridge_weight -= out

        if trucks and bridge_weight + trucks[0] <= weight:
            t = trucks.popleft()
            bridge.append(t)
            bridge_weight += t
        else:
            bridge.append(0)

    return answer